File: src/Pcl.NET/VectorXYZ.cs
using System.Runtime.CompilerServices;

namespace Pcl.NET
{
    public class VectorXYZ : Vector<PointXYZ>
    {
        public VectorXYZ()
        {
            _ptr = Invoke.std_vector_xyz_ctor();
        }

        public VectorXYZ(long count)
        {
            _ptr = Invoke.std_vector_xyz_ctor_count((ulong)count);
        }

        /// <summary>
        /// construct as a copy of the list
        /// </summary>
        /// <param name="list"></param>
        public unsafe VectorXYZ(PointXYZ[] arr)
        {
            _ptr = Invoke.std_vector_xyz_ctor_count((ulong)arr.Length);
            CopyFromArray(arr, DataU);
        }

        internal VectorXYZ(IntPtr ptr)
        {
            _suppressDispose = true;
            _ptr = ptr;
        }

        public override long Count
        {
            get
            {
                ThrowIfDisposed();
                return (long)Invoke.std_vector_xyz_size(_ptr);
            }
        }

        public override IntPtr Data
        {
            get
            {
                return Invoke.std_vector_xyz_data(_ptr);
            }
        }

        public override void Add(PointXYZ item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_xyz_add(_ptr, item);
        }

        public override void At(long index, ref PointXYZ value)
        {
            ThrowIfDisposed();
            if ((ulong)index >= (ulong)Count)
            {
                ThrowHelper.ThrowArgumentOutOfRange_IndexMustBeLessException();
            }

            Invoke.std_vector_xyz_at(_ptr, (ulong)index, ref value);
        }

        public override void Clear()
        {
            ThrowIfDisposed();
            Invoke.std_vector_xyz_clear(_ptr);
        }

        public override void Insert(long index, PointXYZ item)
        {
            ThrowIfDisposed();
            if ((ulong)index >= (ulong)Count)
            {
                ThrowHelper.ThrowArgumentOutOfRange_IndexMustBeLessException();
            }

            Invoke.std_vector_xyz_insert(_ptr, (ulong)index, item);
        }

        public override void Resize(long size)
        {
            ThrowIfDisposed();
            Invoke.std_vector_xyz_resize(_ptr, (ulong)size);
        }

        public override IEnumerator<PointXYZ> GetEnumerator()
        {
            long count = Count;
            for (var i = 0; i < count; i++)
            {
                yield return this[i];
            }
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.std_vector_xyz_delete(ref _ptr);
            }
        }
    }
}

File: src/pcl-bindings/VectorXYZ.cpp
#include "export.h"
#include "pcl\point_types.h"

using namespace std;
using namespace pcl;
using point_t = PointXYZ;
using vector_t = vector<point_t>;

EXPORT(vector_t*) std_vector_xyz_ctor()
{
	return new vector_t();
}

EXPORT(vector_t*) std_vector_xyz_ctor_count(size_t count)
{
    return new vector_t(count);
}

EXPORT(void) std_vector_xyz_delete(vector_t** ptr)
{
	delete* ptr;
	*ptr = NULL;
}

EXPORT(void) std_vector_xyz_at(vector_t* ptr, size_t idx, point_t* value)
{
	*value = ptr->at(idx);
}

EXPORT(size_t) std_vector_xyz_size(vector_t* ptr)
{
	return ptr->size();
}

EXPORT(void) std_vector_xyz_clear(vector_t* ptr)
{
	ptr->clear();
}

EXPORT(void) std_vector_xyz_resize(vector_t* ptr, size_t size)
{
	ptr->resize(size);
}

EXPORT(void) std_vector_xyz_add(vector_t* ptr, point_t value)
{
	//the value needs to be aligned to be pushed into the cloud, so do that.
    point_t deref;
	memcpy(&deref, &value, sizeof(point_t));
	ptr->push_back(deref);
}

EXPORT(void) std_vector_xyz_insert(vector_t* ptr, ptrdiff_t index, point_t value)
{
	//the value needs to be aligned to be pushed into the cloud, so do that.
    point_t deref;
	memcpy(&deref, &value, sizeof(point_t));
	ptr->insert(ptr->begin() + index, deref);
}

EXPORT(point_t*) std_vector_xyz_data(vector_t* ptr)
{
	return ptr->data();
}